mlir/lib/Dialect/LLVMIR/IR/LLVMInlining.cpp

```diff
@@ -50,11 +50,29 @@ static bool hasLifetimeMarkers(LLVM::AllocaOp allocaOp) {
 static void
 handleInlinedAllocas(Operation *call,
                      iterator_range<Region::iterator> inlinedBlocks) {
+  // The traversal's objective is to locate the entry block of the closest call
+  // site ancestor that has the automatic allocation scope trait. In pure LLVM
+  // dialect programs, this is the LLVM function containing the call site.
+  // However, mixed dialect programs may have other allocation scopes.
+  assert(call->getParentWithTrait<OpTrait::AutomaticAllocationScope>() &&
+         "expected call site to be in an automatic allocation scope");
+  Block *allocaScopeEntryBlock = nullptr;
+  Operation *currentOp = call;
+  while (Operation *parentOp = currentOp->getParentOp()) {
+    if (parentOp->hasTrait<OpTrait::AutomaticAllocationScope>()) {
+      allocaScopeEntryBlock = &currentOp->getParentRegion()->front();
+      break;
+    }
+    currentOp = parentOp;
+  }
```

Comment:
What if there is an isolatedFromAbove operation along the way?

Reply:
I would assume that an operation that is closed from above also has the automatic allocation scope trait? I can definitely add more checks the question is a bit if this is a problem we should rather solve downstream?  There can always be an operation with unknown semantics and being conservative here may make more sense (aka closing this revision)?

WDYT?

Reply:
> I would assume that an operation that is closed from above also has the automatic allocation scope trait? 

Maybe, but if it does not I would be concerned that you'd generated invalid IR.

Reply:
Yes I agree there is no trait that guarantees we can hoist an alloca in the general case. Isolated from above guarantees that the value is visible. However, the region may have parallel semantics or other unknown semantics that makes hoisting invalid. We will solve the problem downstream.

Reply:
>  However, the region may have parallel semantics or other unknown semantics that makes hoisting invalid. We will solve the problem downstream.

Such a region should have automatic allocation trait and/or be isolatedFromAbove.
I would think hoisting should be legal otherwise.

Reply:
Ok I tried fix using `mightHaveTrait` and checking for both traits.